Add black support as a v2 rule

pierrechevalier83 opened the pull request:
### Problem

We want to provide a way for pants to run black.

### Solution

Add a v2 rule for "fmt_v2" (temporarily, to workaround conflict with the v1 "fmt" rule).
Add a run_black rule which runs black on each target.
Register Black's config file as a configurable option.
Add a black config that works for us (line length 100 lines which is aligned with our current style,
exclude the file in "compilation_failure" as it is intentionally invalid and trips black up).

### Result

It is now possible to run `./pants fmt_v2 ::` to format all python files in the repository (or any other target, as expected).

Note that this isn't the full picture, but it is probably good enough to stand as its own PR.
Things we may want to improve in the future:
* Make it part of `./pants fmt` (and skip the python formatting by default to avoid surprising our users)
* Minimise the calls to Black by batching all the files form all targets being run and calling it only once
  This will yield speed improvement and a less verbose command line output.
* Add it as part of our pre_commit hook.
* Format our code with it.

pierrechevalier83 commented:
Note: this is one of the prerequisites for https://github.com/pantsbuild/pants/issues/8334

illicitonion commented on pyproject.toml:
```diff
@@ -0,0 +1,14 @@
+line-length = 100
+exclude = '''
+/(
+  | \.git
```
From a pants perspective, pants already ignores these files when it expands globs; are they here in order to preserve the ability to just run `black` not via pants? If so, may be worth a comment to that effect

illicitonion commented on src/python/pants/backend/python/rules/python_fmt.py:
```diff
@@ -0,0 +1,128 @@
+from pants.util.objects import datatype
+
+
+class PythonFormatable(datatype(['target'])):
```
Maybe `FormattablePythonTarget`?

illicitonion replied:
Also, please add a link to https://github.com/pantsbuild/pants/issues/8343 explaining that this is a horrible hack, and why it's there :)

illicitonion commented on src/python/pants/backend/python/rules/python_fmt.py:
```diff
@@ -0,0 +1,128 @@
+    RequirementsPex, RequirementsPexRequest(
+      output_filename="black.pex",
+      requirements=tuple(black.get_requirement_specs()),
+      interpreter_constraints=(),
```
`PythonToolBase` by default has a flag for this, so we should probably pass it through from the `Black` instance.

Also, IIRC black has a minimum python version of 3.6, so we probably want `Black` to override `default_interpreter_constraints` to specify that minimum.

pierrechevalier83 replied:
> Also, IIRC black has a minimum python version of 3.6, so we probably want Black to override default_interpreter_constraints to specify that minimum.
By default, it picks the flag based on the file itself, so I don't think that's necessary.

illicitonion replied:
I'm not sure I follow... What would stop this pex from trying to run with py27 if that's the default python on your system?

pierrechevalier83 replied:
Sorry: misread your comment. I thought you were speaking about Black's option of specifying the version of the file that it's formatting. Addressed now.

illicitonion commented on src/python/pants/backend/python/rules/python_fmt.py:
```diff
@@ -0,0 +1,128 @@
+      entry_point=black.get_entry_point(),
+    )
+  )
+  target = target.target
```
Maybe call the parameter `wrapped_target`, so this line reads `target = wrapped_target.target`?

illicitonion commented on src/python/pants/backend/python/rules/python_fmt.py:
```diff
@@ -0,0 +1,128 @@
+  # The exclude option from Black only works on recursive invocations,
+  # so call black with the directories in which the files are present
+  # and passing the full file names with the include option
+  dirs = []
```
This should probably be a `set`, rather than a list; multiple files may be in the same dir

(and then you'll want `tuple(sorted(dirs))` when you consume them)

illicitonion commented on src/python/pants/backend/python/rules/python_fmt.py:
```diff
@@ -0,0 +1,128 @@
+  if config_path:
+    pex_args += ("--config", config_path)
+  if target.sources.snapshot.files:
+    pex_args += ("--include",) + target.sources.snapshot.files
```
From the docs of `black`, I'm not quite sure what this is doing. `--include` takes a single argument, but this is specifying many. Can we remove the `--include` and it will just format things in the listed directories? Or do we need to regexify the list of files into a single argument?

illicitonion commented on src/python/pants/rules/core/fmt.py:
```diff
@@ -0,0 +1,64 @@
+
+@union
+class FmtTarget:
+  """A union for registration of a testable target type."""
```
s/testable/formattable/

illicitonion commented on src/python/pants/rules/core/fmt.py:
```diff
@@ -0,0 +1,64 @@
+class Fmt(Goal):
+  """Autoformat source code."""
+
+  name = 'fmt_v2'
```
Add a TODO linking to https://github.com/pantsbuild/pants/issues/8351

illicitonion commented on src/python/pants/rules/core/fmt.py:
```diff
@@ -0,0 +1,64 @@
+
+@console_rule
+def fmt(console: Console, targets: HydratedTargets) -> Fmt:
+  results = yield [
```
Add a TODO linking to https://github.com/pantsbuild/pants/issues/4535 saying something like "`@union` assumes that all targets passed implement the union, so we manually filter the targets we know do; this should probably no-op or log or something configurable for non-matching targets"

illicitonion commented on src/python/pants/rules/core/fmt.py:
```diff
@@ -0,0 +1,64 @@
+  results = yield [
+          Get(FmtResult, FmtTarget, target.adaptor)
+          for target in targets
+          if isinstance(target.adaptor, (PythonAppAdaptor, PythonTargetAdaptor, PythonTestsAdaptor, PythonBinaryAdaptor)) and hasattr(target.adaptor, "sources")
```
Add a TODO linking to https://github.com/pantsbuild/pants/issues/4535 complaining about the `hasattr` check we need to do here

illicitonion commented on src/python/pants/rules/core/fmt.py:
```diff
@@ -0,0 +1,64 @@
+    files_content = yield Get(FilesContent, Digest, result.digest)
+    for file_content in files_content:
+      with open(os.path.join(get_buildroot(), file_content.path), "wb") as f:
+        f.write(file_content.content)
```
Add a TODO that we should use https://github.com/pantsbuild/pants/pull/8329 (and noting that the reason we're ok with this horribly inefficient implementation is because it will disappear soon :D)

illicitonion commented on src/python/pants/rules/core/fmt.py:
```diff
@@ -0,0 +1,64 @@
+      with open(os.path.join(get_buildroot(), file_content.path), "wb") as f:
+        f.write(file_content.content)
+
+    console.print_stdout(result.stdout)
```
I _think_ these will print newlines if there's no stdout/stderr, so let's either wrap these in an `if result.std*:` check, or use `console.write_std*` to avoid those extra newlines. Or pass `end=''`.

pierrechevalier83 commented:
I addressed some of the more minor comments and fixed a bug Daniel spotted. Making integration tests will take me a bit longer, so I'll add it to this PR later.

illicitonion commented:
> Not deeply reviewing right now, but I think that #8329, and maybe _some_ prework for #4535 should be blockers here? I'll add a comment to #4535 with something that we could potentially break out (edit: [here](https://github.com/pantsbuild/pants/issues/4535#issuecomment-536075252)).

The hacks to work around those issues are pretty well localised, and this is a default-off goal (it doesn't shadow any existing goal namespace), so I think we should be good to merge this when it has some tests and do some incremental improvements as the other issues get resolved, but happy to discuss more if you disagree :)

pierrechevalier83 commented:
Added some integration tests

illicitonion commented on tests/python/pants_test/backend/python/rules/test_python_fmt_integration.py:
```diff
@@ -0,0 +1,70 @@
+from pants.util.contextutil import temporary_file_path
+from os.path import relpath
+
+class PythonFmtIntegrationTest(PantsRunIntegrationTest):
```
Can we add one test which _does_ do formatting, e.g. by creating a temporary file in-line in the test?

Eric-Arellano commented on src/python/pants/backend/python/rules/python_fmt.py:
```diff
@@ -0,0 +1,137 @@
+  # and passing the full file names with the include option
+  dirs = set()
+  for filename in target.sources.snapshot.files:
+    dirs.add(os.path.dirname(filename))
```
Instead, use `pathlib.Path`! It's overall a much nicer API.

```python
from pathlib import Path

dirs.add(Path(filename).parent)
```

Eric-Arellano replied:
Btw to go from `Path` -> `str`, call `str(my_path)` or `f"{my_path}"`.

Eric-Arellano commented on src/python/pants/backend/python/rules/python_fmt.py:
```diff
@@ -0,0 +1,137 @@
+  # The exclude option from Black only works on recursive invocations,
+  # so call black with the directories in which the files are present
+  # and passing the full file names with the include option
+  dirs = set()
```
Empty collections like this should be typed:

```python
from typing import Set

dirs: Set[Path] = set()
```

(Path is if you take the below recommendation. Otherwise, it's `Set[str]`.

Eric-Arellano commented on src/python/pants/backend/python/rules/python_fmt.py:
```diff
@@ -0,0 +1,137 @@
+  if config_path:
+    pex_args += ("--config", config_path)
+  if target.sources.snapshot.files:
+    pex_args += ("--include", "|".join(re.escape(file) for file in target.sources.snapshot.files))
```
`file` is a builtin. Instead, call it `f` or `fp`.

Eric-Arellano commented on src/python/pants/rules/core/fmt.py:
```diff
@@ -0,0 +1,81 @@
+  ('digest', Digest),
+  ('stdout', str),
+  ('stderr', str),
+])):
```
Conver to dataclass

```python
from dataclasses import dataclass

@dataclass(frozen=True)
class FmtResult:
  digest: Digest
  stdout: str
  stderr: str
```

Make sure you include `3rdparty/python:dataclasses` in BUILD.

Eric-Arellano commented on src/python/pants/rules/core/fmt.py:
```diff
@@ -0,0 +1,81 @@
+    # Blocked on: https://github.com/pantsbuild/pants/pull/8329
+    for file_content in files_content:
+      with open(os.path.join(get_buildroot(), file_content.path), "wb") as f:
+        f.write(file_content.content)
```
```python
from pathlib import Path

Path(get_buildroot(), file_content.path).write_text(file_content.content.decode())
```

Eric-Arellano replied:
Or

```python
from pathlib import Path

with Path(get_buildroot(), file_content.path).open('wb') as f:
  f.write(file_content.content)
```

Eric-Arellano commented on src/python/pants/rules/core/fmt.py:
```diff
@@ -0,0 +1,81 @@
+    if result.stderr:
+      console.print_stderr(result.stderr)
+
+  # workspace.materialize_directories(tuple(digests))
```
Leftover?

Eric-Arellano commented:
Also when testing your new tests, be sure to run `./pants --no-v1 --v2 test`, which is what CI uses by default.

stuhood commented on src/python/pants/backend/python/rules/python_fmt.py:
```diff
@@ -0,0 +1,137 @@
+  ]
+  merged_input_files = yield Get(
+    Digest,
+    DirectoriesToMerge,
```
Can use the two-arg form of `Get` in this case.

stuhood commented on src/python/pants/backend/python/rules/python_fmt.py:
```diff
@@ -0,0 +1,137 @@
+  )
+
+
+@rule
```
Please add a TODO on these definitions pointing to the ticket(s) you opened about unions.

stuhood commented on src/python/pants/rules/core/fmt.py:
```diff
@@ -0,0 +1,81 @@
+      with open(os.path.join(get_buildroot(), file_content.path), "wb") as f:
+        f.write(file_content.content)
+
+    if result.stdout:
```
This feels likely to be noisy... but I suppose that if a tool knows that it won't produce anything interesting, it could skip including it?

pierrechevalier83 replied:
I think that can be made less verbose by batching the files later. I would probably not discard the output as I would want to see what happened as a user.

pierrechevalier83 commented:
https://travis-ci.org/pantsbuild/pants/jobs/592156389
So `./pants reference` failed. That's something we may want to do as part of the pre-commit hook in the future to help catch these kind of errors earlier.
Fixed for now in an extra commit.

Eric-Arellano commented on src/python/pants/backend/python/rules/python_fmt.py:
```diff
@@ -0,0 +1,141 @@
+# containing python files so we can have one single type used as an input in the run_black rule.
+class FormattablePythonTarget(datatype(['target'])):
+  pass
+
```
Didn't notice this in last review. This should also be a dataclass instead of datatype. We want to use dataclass as much as possible now.

Eric-Arellano replied:
Bump on removing `datatype`.

Eric-Arellano commented on src/python/pants/backend/python/rules/python_fmt.py:
```diff
@@ -0,0 +1,141 @@
+  # and passing the full file names with the include option
+  dirs: Set[str] = set()
+  for filename in target.sources.snapshot.files:
+    dirs.add(f"{Path(filename).parent}")
```
Thanks for these changes :)

Eric-Arellano commented on src/python/pants/backend/python/rules/python_fmt.py:
```diff
@@ -0,0 +1,141 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import os
```
I think this is unused

Eric-Arellano commented on src/python/pants/rules/core/fmt.py:
```diff
@@ -0,0 +1,80 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import os
```
Unused I think

Eric-Arellano commented on tests/python/pants_test/backend/python/rules/test_python_fmt_integration.py:
```diff
@@ -0,0 +1,78 @@
+        ]
+      pants_run = self.run_pants(command=command)
+      code = open(file_name)
+      formatted = code.read();
```
Instead:
```python
formatted = Path(file_name).read_text()
```

Has the additional benefit of closing the file for you.

Eric-Arellano commented on tests/python/pants_test/backend/python/rules/test_python_fmt_integration.py:
```diff
@@ -0,0 +1,78 @@
+      pants_run = self.run_pants(command=command)
+      code = open(file_name)
+      formatted = code.read();
+      self.assertEqual("x = 42\n", formatted)
```
Totally a style thing: I like to dedent this to put it outside the context manager (with statement). Now that `formatted` has the value it needs saved, this assertion can move outside.

I try to keep the code within `with` statements as minimum as possible - only have the code that requires that context.

--

Another related idiom in Python is `try-except-else`:

```
try:
  run_command_that_may_raise_an_exception()
except:
  handle_error()
else:  # if there was no exception
  run_the_rest_of_the_code()
```

Usually the `try` block has only 1-2 lines and most of the logic goes into `else`. This makes it really explicit what code might be raising the error.

Eric-Arellano commented on tests/python/pants_test/pants_run_integration_test.py:
```diff
@@ -327,6 +328,8 @@ def run_pants_with_workdir_without_waiting(self, command, workdir, config=None,
         ini.write(fp)
       args.append('--pants-config-files=' + ini_file_name)
 
+    Path('pyproject.toml').touch()
```
Instead of this, add `//:pyproject` to the requirements for `pants.pex` here:

https://github.com/pantsbuild/pants/blob/f673fd2a4f59ea310f24ed702a01779118dad88f/src/python/pants/bin/BUILD#L71-L90

pierrechevalier83 replied:
I had tried this better and it didn't work because the buildroot is not guarenteed to be the same. Fixing to take your suggestion into account and do "the right thing"

Eric-Arellano commented on tests/python/pants_test/rules/BUILD:
```diff
@@ -38,3 +38,16 @@ python_tests(
+  source='test_fmt_integration.py',
+  dependencies=[
+    'tests/python/pants_test:int-test',
+    '//:pyproject',
```
Remove this because of the above recommendation.

Eric-Arellano commented on tests/python/pants_test/rules/BUILD:
```diff
@@ -38,3 +38,16 @@ python_tests(
+    'testprojects/tests/java/org/pantsbuild/testproject:dummies_directory',
+  ],
+  tags={'integration'},
+  timeout = 90,
```
Why put this? This one IT probably shouldn't take 90 seconds.

Eric-Arellano commented on tests/python/pants_test/backend/python/rules/BUILD:
```diff
@@ -40,3 +40,16 @@ python_tests(
+    'examples/src/python/example:hello_directory',
+  ],
+  tags={'integration'},
+  timeout = 90,
```
How long does this take to run locally?

pierrechevalier83 replied:
Yeah, that's arbitrary. I thought 90 seconds was an OK default, but it should take less than 10 seconds. Just removed it and kept the default.

Eric-Arellano commented on tests/python/pants_test/rules/test_fmt_integration.py:
```diff
@@ -0,0 +1,16 @@
+
+
+class FmtIntegrationTest(PantsRunIntegrationTest):
+  def test_black_no_python_sources_should_noop(self):
```
I'd generify this name given that it's in the core `rules` folder. Something like `test_noop_when_no_fmt_available_for_target_type`

pierrechevalier83 commented:
Addressed the latest comments. I *think* CI should also be good now (except for spurious issues)

Eric-Arellano commented on tests/python/pants_test/backend/python/rules/test_python_fmt_integration.py:
```diff
@@ -0,0 +1,78 @@
+        'examples/src/python/example/hello/greet:greet'
+        ]
+      pants_run = self.run_pants(command=command)
+      formatted = Path(file_name).read_text();
```
Unnecessary semicolon. I think Black will remove this in a bit so probably not worth a CI burn.

Eric-Arellano commented on tests/python/pants_test/pants_run_integration_test.py:
```diff
@@ -327,6 +328,9 @@ def run_pants_with_workdir_without_waiting(self, command, workdir, config=None,
       args.append('--pants-config-files=' + ini_file_name)
 
+    if not Path(get_buildroot()).samefile(Path(".")):
+      shutil.copy(str(Path(get_buildroot(), "pyproject.toml")), str(Path(".")))
```
Hm this is surprising and concerning to me. Are you sure it’s necessary? We don’t have to do this for any other tooling config files, and the change you made means that the pants.pex will already have the config in its chrooted folder.

pierrechevalier83 replied:
I'm sure it's necessary. It tripped me up for a good part of the morning because I had the same expectation you had. That's why I ended up just touching the file, but I still think this latest version is more correct.

Eric-Arellano replied:
What happens if you include it as a dependency on the Pants.pex binary target like you have now AND the IT target? Not sure it will make a difference, but possibly. 

I gotta run to class and don’t want to block on this because we can always address in a followup, but would love to make sure there’s no way to avoid this.

pierrechevalier83 replied:
OK. Found how it works for pants.pex etc. Leads me to a slightly more elegant solution. Will push new commit. (Still a hack, but the same as others)

pierrechevalier83 replied:
See my latest commit (hopefully explains what's going on)
